## Collaborative Filtering - MovieLens Dataset

Author: Khushee Kapoor

Last Updated: 26/12/2021

Week: 6

#### **Q1.** Read about the movielens datasetand write down a summary of metadata.

**Meta-Data** of the MovieLens Dataset:

The data is developed by GroupLens Research for their Movie Recommender System Project - MovieLens. It contains four csv datasets - movies.csv, ratings.csv, links.csv and tags.csv each of which have either of the two identifiers - userId and movieId. The dataset was developed over a period of 22 years wherein only those randomized and anonymized users who have rated atleast 20 movies have been selected, and only those movies which have atleast 1 rating or tag have been selected. Description of the columns in the dataset:

Identifiers:

*  userId - random and anonymous IDs given to identify users. This is the only way a user can be recognized as no other demographic information is given. The IDs are a complete sequence from 1 to 610.
*   movieId - IDs given to identify movies. Since only those movies which have 1 or more rating or tag are selected, it is not a complete sequence.

movies.csv:
*   title -  the title of the movie alongwith the year of release.
*   genre - tags of genres given. Possible values: action, adventure, animation, children's, comedy, crime, documentary, drama, fantasy, film-noir, horror, musical, mystery, romance, sci-fi, thriller, war, western, (no genre listed).

ratings.csv:
*   rating - ratings given by user. Maximum Value: 5; Minimum Value: 0.5; Step Value: 0.5.
*   timestamp - timestamp of when the movie was rated (UTC Time).

links.csv:
*   imdbId: link to the imdb page of the movie.
*   tmbdId: link to the moviedb page of the movie.

tags.csv:
*   tag: a word or short phrase that describes the user's impressions about the movie. 
*   timestamp - timestamp of when the movie was rated (UTC Time).


The datases have been ordered by the userIds, and inside the userId, by the movieId.






First, we import the libraries:

*  NumPy - for data manipulation.
*  Pandas - for data manipulation.
*  MatPlotLib - for data visualization.
*  Seaborn - for data visualization.

In [1]:
# importing the libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

Next, we read all the datasets and store them in a Pandas dataframe. To do this, we use the read_csv() function from the Pandas library. We also view the first few rows of each dataset to get a glimpse.

In [2]:
# reading the movies dataset
movies = pd.read_csv('movies.csv')

In [3]:
# viewing the first few rows of the movies dataset
movies.head()

movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [4]:
# reading the ratings dataset
ratings = pd.read_csv('ratings.csv')

In [5]:
# viewing the first few rows of the ratings dataset
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [6]:
# reading the links dataset
links = pd.read_csv('links.csv')

In [7]:
# viewing the first few rows of the links dataset
links.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [8]:
# reading the tags dataset
tags = pd.read_csv('tags.csv')

In [9]:
# viewing the first few rows of the tags dataset
tags.head()

userId  movieId              tag   timestamp
0       2    60756            funny  1445714994
1       2    60756  Highly quotable  1445714996
2       2    60756     will ferrell  1445714992
3       2    89774     Boxing story  1445715207
4       2    89774              MMA  1445715200

#### **Q2.** Read the “ratings.csv” file and create a pivot table with index=‘userId’, columns=‘movieId’, values = “rating". 

To solve Question 2, we use the pivot_table() function from the Pandas library, and pass the values 'userId', 'movieId', and 'rating' to the parameters index, columns, and values respectively. This way, every individual row represents a user and columns represent movies.

In [10]:
# creating a pivot table
pd.pivot_table(ratings, index='userId', columns='movieId', values='rating')

movieId  1       2       3       4       ...  193583  193585  193587  193609
userId                                   ...                                
1           4.0     NaN     4.0     NaN  ...     NaN     NaN     NaN     NaN
2           NaN     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
3           NaN     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
4           NaN     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
5           4.0     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
...         ...     ...     ...     ...  ...     ...     ...     ...     ...
606         2.5     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
607         4.0     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
608         2.5     2.0     2.0     NaN  ...     NaN     NaN     NaN     NaN
609         3.0     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN
610         5.0     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN

[610 rows x 9724 columns]

As we can see, there are a lot of NaN or missing values in the dataset. To handle the missing values, we will fill them with a constant. The constant has to be a value that does not clash with the values in the dataset. To do so, let's find the ratings present in the dataset.

In [11]:
# finding all the values in the rating column
ratings.rating.unique()

array([4. , 5. , 3. , 2. , 1. , 4.5, 3.5, 2.5, 0.5, 1.5])

As we can see, no user has given a rating of '0' to any movie. Hence, we use the constant 0 to fill in the missing values by setting the parameter fill_value to 0, and convert it to a Pandas DataFrame using the DataFrame() function from the Pandas library for further analysis.

In [12]:
# filling the missing values in the pivot table with 0 and converting to DataFrame
df = pd.DataFrame(pd.pivot_table(ratings, index='userId', columns='movieId', values='rating', fill_value=0))
df

movieId  1       2       3       4       ...  193583  193585  193587  193609
userId                                   ...                                
1           4.0     0.0     4.0     0.0  ...     0.0     0.0     0.0       0
2           0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
3           0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
4           0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
5           4.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
...         ...     ...     ...     ...  ...     ...     ...     ...     ...
606         2.5     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
607         4.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
608         2.5     2.0     2.0     0.0  ...     0.0     0.0     0.0       0
609         3.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0
610         5.0     0.0     0.0     0.0  ...     0.0     0.0     0.0       0

[610 rows x 9724 columns]

As we can see, the missing values are now filled with 0. Hence, a rating more than 0 represents that the user has watched the movie and given the rating, and the value 0 represents that the user has not watched the movie. Another reason for filling the missing values with 0 was to make the pivot table numeric for further analysis. Also, the last userId is same as the number of rows, meaning no missing values in the sequence.

#### **Q3, 4, 5.** Importing the required packages.

To solve further Questions, we download the following packages to compute the distances and hence similarity between the users.

In [13]:
# importing the packages
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation

#### **Q6.** Find the 5 most similar user for user with user Id 10.

To solve Question 6, we first compute pairwise distances between the rows, that is the users, based on the cosine similarity. To do this, we use the pairwise_distances() function from the ScikitLearn library and set the value of the parameter metric to 'cosine'.

In [14]:
# computing pairwise distances using cosine similarity between the users
distance_matrix = pairwise_distances(df, metric='cosine')

Next, we splice and select the 9th row from the resultant distance matrix to select the 10th User. We do this because arrays in Python are 0-indexed and the UserIDs start from 1. Post that, we sort the cosine similarities in the descending order to find the highest cosine similarity between users, and hence the most similar users. We use the argsort() function on the array to return the indices, and hence the UserIDs of the sorted array. Then, we splice the array to find the top 5 similar users. To do this, we splice values from 1 to 6. We leave out the index 0 because it is by default the same user. After splicing, we add 1 to the indices since arrays in Python are 0-indexed and UserIDs start from 1. 

In [15]:
# finding the top 5 most similar users
(distance_matrix[9]).argsort()[1:6]+1

array([159, 143, 563, 177, 189])

As we can see, the users 159, 143, 563, 177 and 189 are most similar to user 10.

#### **Q7.** Use the “movies”dataset to find out the names of movies, user 2 and user 338 have watched in common and how they have rated each one of them.

To solve Question 7, first we use the loc[] function from the Pandas library and filter the dataset to get movieIds rated by Users 2 and 338. Then, we pass both of these filtered dataframes to the set() function to form sets and then find the intersection of these sets, hence, finding the movies that both the users have watched in common.

In [16]:
# finding the intersection of the sets of the movies that both users 2 and 338 have watched
set(ratings.loc[ratings.userId==2, 'movieId']).intersection(set(ratings.loc[ratings.userId==338, 'movieId']))

{318, 6874}

As we can see, movies having movieIds 318 and 6874 are the ones watched by both the users.

Next, we use the loc[] function from the Pandas library to find the title and rating of the movies in common watched by both the users.

In [17]:
# printing the ratings given by users 
print('Movie ID 318: ', movies.loc[movies.movieId==318, 'title'].values[0])
print('Rating by User 2: ', ratings.loc[((ratings.userId==2) & (ratings.movieId==318)), 'rating'].values[0])
print('Rating by User 338: ', ratings.loc[((ratings.userId==338) & (ratings.movieId==318)), 'rating'].values[0])
print()
print('Movie ID 6874: ', movies.loc[movies.movieId==6874, 'title'].values[0])
print('Rating by User 2: ', ratings.loc[((ratings.userId==2) & (ratings.movieId==6874)), 'rating'].values[0])
print('Rating by User 338: ', ratings.loc[((ratings.userId==338) & (ratings.movieId==6874)), 'rating'].values[0])

Movie ID 318:  Shawshank Redemption, The (1994)
Rating by User 2:  3.0
Rating by User 338:  5.0

Movie ID 6874:  Kill Bill: Vol. 1 (2003)
Rating by User 2:  4.0
Rating by User 338:  4.5


As we can see, the movie having movieId 318 or The Shawshank Redemption was given a rating of 3.0 by User 2 and 5.0 by User 338, whereas the movie having movieId 6874 or Kill Bill Vol. 1 was given a rating of 4.0 by User 2 and 4.5 by User 338. This indicates that User 2 is more critical in watching movies.

#### **Q8.** Use the movies dataset to find out the common movienames between user 2 and user 338 with least rating of 4.0

To solve Question 8, first we use the loc[] function from the Pandas library and filter the dataset to get movieIds rated by Users 2 and 338 which have rating more than or equal to 4.0. Then, we pass both of these filtered dataframes to the set() function to form sets and then find the intersection of these sets, hence, finding the movies that both the users have watched in common, having rating more than or equal to 4.0.

In [18]:
# finding the intersection of the sets of the movies that both users 2 and 338 have watched and rating more than or equal to 4.0
set(ratings.loc[((ratings.userId==2) & (ratings.rating>=4.0)), 'movieId']).intersection(set(ratings.loc[((ratings.userId==338) & (ratings.rating>=4.0)), 'movieId']))

{6874}

As we can see, movie having movieId 6874 is common between users 2 and 338 and has rating more than or equal to 4.0.

Next, we use the loc[] function from the Pandas library to find the title of the movies in common watched by both the users and having rating more than or equal to 4.0.

In [19]:
# printing the title of the movie
print('Movie ID 6874: ', movies.loc[movies.movieId==6874, 'title'].values[0])

Movie ID 6874:  Kill Bill: Vol. 1 (2003)


As we can see, movie having movieId 6874 is Kill Bill Vol. 1.

#### **Q9.** Create a pivot table for representing the similarity among movies using correlation.

To solve Question 9, we first merge the dataframes ratings and movies using the merge() function from the Pandas library on the column movieId.

In [20]:
# merging the dataframes ratings and movies
df = ratings.merge(movies, on='movieId')

Next, we use the pivot_table() function from the Pandas library, and pass the values 'title', 'userId', and 'rating' to the parameters index, columns, and values respectively. As from our analysis in Question 2, we fill the missing values with 0, and convert it to a Pandas DataFrame using the DataFrame() function from the Pandas library for further analysis.

In [21]:
# creating the pivot table and convert it to a DataFrame
df = pd.DataFrame(pd.pivot_table(df, index='title', columns='userId', values='rating', fill_value=0))
df

userId                                     1    2    3    ...  608  609  610
title                                                     ...               
'71 (2014)                                   0  0.0  0.0  ...  0.0    0  4.0
'Hellboy': The Seeds of Creation (2004)      0  0.0  0.0  ...  0.0    0  0.0
'Round Midnight (1986)                       0  0.0  0.0  ...  0.0    0  0.0
'Salem's Lot (2004)                          0  0.0  0.0  ...  0.0    0  0.0
'Til There Was You (1997)                    0  0.0  0.0  ...  0.0    0  0.0
...                                        ...  ...  ...  ...  ...  ...  ...
eXistenZ (1999)                              0  0.0  0.0  ...  4.5    0  0.0
xXx (2002)                                   0  0.0  0.0  ...  3.5    0  2.0
xXx: State of the Union (2005)               0  0.0  0.0  ...  0.0    0  1.5
¡Three Amigos! (1986)                        4  0.0  0.0  ...  0.0    0  0.0
À nous la liberté (Freedom for Us) (1931)    0  0.0  0.0  ...  0.0    0  0.0

[9719 rows x 610 columns]

As we can see from the pivot table above, the last movieId exceeds the number of rows. That means that the movieId and the row numbers are not the same. This is because only those movies who have atleast 1 rating or tag have been chosen.

#### **Q10.** Find the top 5  movies which are similar to the movie “Godfather”.

To solve Question 10, we first compute pairwise distances between the rows, that is the movies, based on the correlation. To do this, we use the pairwise_distances() function from the ScikitLearn library and set the value of the parameter metric to 'correlation'.

In [22]:
# computing pairwise distances using correlation between the movies
distance_matrix = pairwise_distances(df, metric='correlation')

Next, we find the movies which contain the word Godfather. To do this, we use string functions and filter the dataframe.

In [23]:
# finding the movies which contain the word Godfather
df[df.index.str.contains('Godfather')]

userId                                   1    2    3    4    ...  607  608  609  610
title                                                        ...                    
Godfather, The (1972)                      0  0.0  0.0    0  ...    4  5.0    0  5.0
Godfather: Part II, The (1974)             0  0.0  0.0    0  ...    0  4.5    0  5.0
Godfather: Part III, The (1990)            0  0.0  0.0    0  ...    0  4.5    0  0.0
The Godfather Trilogy: 1972-1990 (1992)    0  0.0  0.0    0  ...    0  0.0    0  0.0
Tokyo Godfathers (2003)                    0  0.0  0.0    0  ...    0  0.0    0  0.0

[5 rows x 610 columns]

As we can see, there are 5 movies which contain the word Godfather. However, we choose the first movie - Godfather, The (1972), which is the oldest.

Since the movieId and row number is not the same, we find the index of the Godfather movie. To do this, we first make a list of all the movies and then find the index of the Godfather movie in the list.

In [24]:
# finding index of Godfather movie
indices = list(df.index) 
indices.index('Godfather, The (1972)')

3499

Next, we use the index previously found and splice the distance_matrix. Post that, we sort the correlations in the descending order to find the highest correlation between movies, and hence the most similar movies. We use the argsort() function on the array to return the indices. Then, we splice the array to find the top 5 similar movies. To do this, we splice values from 1 to 6. We leave out the index 0 because it is by default the same movie. Then, we print the movies by using the indices and finding their values in the indices list we previously made.

In [25]:
# finding the top 5 similar movies
similar_indices = (distance_matrix[3499].argsort()[1:6])
for i in similar_indices:
  print(indices[i])

Godfather: Part II, The (1974)
Goodfellas (1990)
One Flew Over the Cuckoo's Nest (1975)
Reservoir Dogs (1992)
Fargo (1996)


As we can see, the movies Godfather: Part II, The (1974), Goodfellas (1990), One Flew Over the Cuckoo's Nest (1975), Reservoir Dogs (1992), and Fargo (1996) are the most similar to Godfather, The (1972).